File: backend/app/models/narration.py
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.config import MAX_TEXT_CHARS

Emotion = Literal[
    "calm",
    "tense",
    "anxious",
    "dramatic",
    "romantic",
    "relieved",
    "excited",
    "sad",
    "angry",
]
VoiceRole = Literal["narrator", "male_character", "female_character"]
Pace = Literal["slow", "medium", "fast"]
DirectionMode = Literal["cinematic", "grounded", "intimate"]


class NarrationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    text: str = Field(..., min_length=1, max_length=MAX_TEXT_CHARS)
    direction_mode: DirectionMode = "cinematic"
    reader_feedback: Optional[str] = None
    previous_emotion: Optional[Emotion] = None

    @field_validator("text")
    @classmethod
    def validate_text(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Text cannot be empty.")
        if len(value) > MAX_TEXT_CHARS:
            raise ValueError(
                f"Text is too long. Keep requests under {MAX_TEXT_CHARS} characters."
            )
        return value

    @field_validator("reader_feedback")
    @classmethod
    def validate_reader_feedback(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None

        trimmed = value.strip()
        if not trimmed:
            return None
        return trimmed[:240]


class NarrationConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scene_type: str
    speaker_role: VoiceRole
    emotion: Emotion
    intensity: int
    pace: Pace
    pause_after_sentence: bool
    audio_tag: Optional[str] = "none"


class CharTimelineItem(BaseModel):
    char: str
    index: int
    start: float
    end: float
    role: VoiceRole


class WordTimelineItem(BaseModel):
    word: str
    start: float
    end: float
    char_start: int
    char_end: int
    role: VoiceRole


class TimelineSegment(BaseModel):
    role: VoiceRole
    emotion: Emotion
    intensity: int
    start: float
    end: float
    text: str
    audio_tag: str = "none"


class NarrationMetadata(BaseModel):
    cache_hit: bool = False
    analysis_cache_hit: bool = False
    segment_count: int
    word_count: int
    duration_seconds: float
    dominant_emotion: Emotion
    default_pace: Pace
    direction_mode: DirectionMode


class NarrationResponse(BaseModel):
    audio: str
    timeline: list[TimelineSegment]
    char_timeline: list[CharTimelineItem]
    word_timeline: list[WordTimelineItem]
    metadata: NarrationMetadata

File: frontend/lib/narration.ts
export type Emotion =
  | "calm"
  | "tense"
  | "anxious"
  | "dramatic"
  | "romantic"
  | "relieved"
  | "excited"
  | "sad"
  | "angry";

export type VoiceRole = "narrator" | "male_character" | "female_character";
export type DirectionMode = "cinematic" | "grounded" | "intimate";

export interface NarrationCharTimelineItem {
  char: string;
  index: number;
  start: number;
  end: number;
  role: VoiceRole;
}

export interface NarrationWordTimelineItem {
  word: string;
  start: number;
  end: number;
  char_start: number;
  char_end: number;
  role: VoiceRole;
}

export interface NarrationTimelineItem {
  role: VoiceRole;
  emotion: Emotion;
  intensity: number;
  start: number;
  end: number;
  text: string;
  audio_tag: string;
}

export interface NarrationMetadata {
  cache_hit: boolean;
  analysis_cache_hit: boolean;
  segment_count: number;
  word_count: number;
  duration_seconds: number;
  dominant_emotion: Emotion;
  default_pace: "slow" | "medium" | "fast";
  direction_mode: DirectionMode;
}

export interface NarrationResponse {
  audio: string;
  timeline: NarrationTimelineItem[];
  char_timeline: NarrationCharTimelineItem[];
  word_timeline: NarrationWordTimelineItem[];
  metadata: NarrationMetadata;
}

export interface RoleTimelineSegment extends NarrationTimelineItem {
  startProgress: number;
  endProgress: number;
}

export interface DirectionOption {
  id: DirectionMode;
  label: string;
  description: string;
}

export interface SampleStory {
  id: string;
  title: string;
  hook: string;
  directionMode: DirectionMode;
  text: string;
}

export const MAX_STORY_CHARS = 12000;

export function normalizeStoryText(text: string) {
  return text.replace(/\r\n/g, "\n").normalize("NFKC");
}

export const DIRECTION_OPTIONS: DirectionOption[] = [
  {
    id: "cinematic",
    label: "Cinematic",
    description: "High contrast, sharper beats, and more dramatic pacing.",
  },
  {
    id: "grounded",
    label: "Grounded",
    description: "Subtle, restrained delivery with realism over spectacle.",
  },
  {
    id: "intimate",
    label: "Intimate",
    description: "Closer, softer performances with emotional nuance.",
  },
];

export const SAMPLE_STORIES: SampleStory[] = [
  {
    id: "lighthouse",
    title: "The Lighthouse",
    hook: "Hope versus inevitability on a cold coast.",
    directionMode: "cinematic",
    text: `The old lighthouse stood against the twilight sky, its beacon dark for the first time in a century.

"We can't let it die," Marcus said, his voice breaking through the silence.

Elena turned to him, her eyes reflecting the dying light. "Some things are meant to end, Marcus. That's just how the world works."

"Not this," he insisted, placing his hand on the cold stone. "Not while I still breathe."

The wind picked up, carrying with it the salt of the sea and the weight of countless stories. They stood there, two souls bound by memory and stubborn hope, refusing to let go of what once was.`,
  },
  {
    id: "alley",
    title: "Rain Alley",
    hook: "A noir exchange that works well for voice switching.",
    directionMode: "grounded",
    text: `Rain fell in silver lines between the buildings, turning the alley into a corridor of broken mirrors.

"You came alone?" Mara asked.

Jonah stopped beneath the flickering sign and looked at her through the downpour. "You told me to."

"I told you to be careful."

He laughed once, tired and hollow. "That would've been a different night."`,
  },
  {
    id: "letter",
    title: "The Letter",
    hook: "A quieter scene built for softer emotional contour.",
    directionMode: "intimate",
    text: `Nora unfolded the letter with both hands, as if too much force might bruise the paper.

It was only three sentences long, but every word seemed to breathe.

"I kept waiting for the right moment to come back," she read aloud. "Then I realized the right moment had already passed."

The room stayed still around her, tender and unforgiving.`,
  },
];
